File: chapter_api_tests/2024-04/validation/test_put_api-v1-projects-{projectRef}-envvars-{env}-{name}.ts
import axios, { AxiosError, AxiosResponse } from 'axios';
import { describe, it, expect, beforeAll } from '@jest/globals';

/**
 * Test suite for PUT /api/v1/projects/{projectRef}/envvars/{env}/{name}
 *
 * This suite covers:
 * 1. Input Validation
 * 2. Response Validation
 * 3. Response Headers Validation
 * 4. Edge Case & Limit Testing
 * 5. Authorization & Authentication Testing
 */

describe('PUT /api/v1/projects/{projectRef}/envvars/{env}/{name}', () => {
  let apiBaseUrl: string;
  let authToken: string;

  const validProjectRef = 'testProject';
  const validEnv = 'development';
  const validName = 'TEST_VAR';

  /**
   * Utility function to build the full endpoint URL.
   */
  const buildUrl = (
    projectRef: string,
    env: string,
    name: string
  ): string => {
    return `${apiBaseUrl}/api/v1/projects/${projectRef}/envvars/${env}/${name}`;
  };

  /**
   * Axios configuration with optional authorization.
   */
  const createAxiosConfig = (token?: string) => {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };

    if (token) {
      headers.Authorization = `Bearer ${token}`;
    }

    return { headers };
  };

  /**
   * Validate the response headers for JSON content type.
   */
  const expectJsonContentType = (response: AxiosResponse) => {
    expect(response.headers['content-type']).toMatch(/application\/json/i);
  };

  beforeAll(() => {
    apiBaseUrl = process.env.API_BASE_URL || 'http://localhost:3000';
    authToken = process.env.API_AUTH_TOKEN || '';
  });

  /**
   * 1) Testing a successful update with valid data.
   */
  it('should update the environment variable successfully with valid data (200)', async () => {
    const url = buildUrl(validProjectRef, validEnv, validName);
    const requestData = {
      value: 'UpdatedValue',
      // Add any other required fields based on the actual schema, e.g. type/secret...
    };

    let response: AxiosResponse;
    try {
      response = await axios.put(url, requestData, createAxiosConfig(authToken));
      expect(response.status).toBe(200);
      expectJsonContentType(response);

      // Example response schema check (Update according to actual SucceedResponse schema)
      // For instance, if SucceedResponse has { success: boolean, message: string }
      expect(response.data).toHaveProperty('success');
      expect(response.data).toHaveProperty('message');
      expect(response.data.success).toBe(true);
    } catch (err: unknown) {
      if (err instanceof AxiosError && err.response) {
        // If this call unexpectedly fails, report details.
        console.error('Unexpected error response:', err.response.data);
      }
      throw err;
    }
  });

  /**
   * 2) Testing invalid request body -> expect 400 or 422.
   */
  it('should return 400 or 422 for invalid request body', async () => {
    const url = buildUrl(validProjectRef, validEnv, validName);

    // Missing required fields or invalid data type.
    const invalidRequestData = {
      // e.g., missing "value" or using an invalid data type
      value: 1234, // Suppose "value" must be a string based on the schema.
    };

    try {
      await axios.put(url, invalidRequestData, createAxiosConfig(authToken));
      // If the request does not fail, then we did not get the expected error.
      throw new Error('Expected 400 or 422 error, but request succeeded.');
    } catch (err: unknown) {
      if (err instanceof AxiosError && err.response) {
        expect([400, 422]).toContain(err.response.status);
        expectJsonContentType(err.response);
        // Example schema check for an error response
        // Could be { error: string, details?: any } depending on actual schema
        expect(err.response.data).toHaveProperty('error');
      } else {
        throw err;
      }
    }
  });

  /**
   * 3) Testing missing or invalid auth token -> expect 401 or 403.
   */
  it('should return 401 or 403 if the auth token is missing or invalid', async () => {
    const url = buildUrl(validProjectRef, validEnv, validName);

    const requestData = {
      value: 'AnyValue',
    };

    try {
      await axios.put(url, requestData, createAxiosConfig('invalid-token'));
      throw new Error('Expected 401 or 403 error, but request succeeded.');
    } catch (err: unknown) {
      if (err instanceof AxiosError && err.response) {
        expect([401, 403]).toContain(err.response.status);
        expectJsonContentType(err.response);
        expect(err.response.data).toHaveProperty('error');
      } else {
        throw err;
      }
    }
  });

  /**
   * 4) Testing resource not found -> expect 404.
   */
  it('should return 404 if the specified project/env/name does not exist', async () => {
    const invalidProjectRef = 'doesNotExist';
    const invalidEnv = 'notARealEnv';
    const invalidName = 'UNKNOWN_VAR';

    const url = buildUrl(invalidProjectRef, invalidEnv, invalidName);
    const requestData = { value: 'AnyValue' };

    try {
      await axios.put(url, requestData, createAxiosConfig(authToken));
      throw new Error('Expected 404 error, but request succeeded.');
    } catch (err: unknown) {
      if (err instanceof AxiosError && err.response) {
        expect(err.response.status).toBe(404);
        expectJsonContentType(err.response);
        expect(err.response.data).toHaveProperty('error');
      } else {
        throw err;
      }
    }
  });

  /**
   * 5) Testing large input data (edge case & limit testing).
   * Provide an extremely long value.
   */
  it('should handle large payloads correctly', async () => {
    const url = buildUrl(validProjectRef, validEnv, validName);

    // 10,000 characters string as an example
    const largeValue = 'X'.repeat(10000);
    const requestData = {
      value: largeValue,
    };

    try {
      const response = await axios.put(url, requestData, createAxiosConfig(authToken));
      // Depending on the API's limits, it may succeed or fail with 400.
      // Adjust expectations as per your API specification.

      // If we expect success:
      expect(response.status).toBe(200);
      expectJsonContentType(response);
      expect(response.data).toHaveProperty('success');
      expect(response.data.success).toBe(true);
    } catch (err: unknown) {
      if (err instanceof AxiosError && err.response) {
        // If there's a size limit, we might get 400 or 413.
        expect([400, 413]).toContain(err.response.status);
        expectJsonContentType(err.response);
        expect(err.response.data).toHaveProperty('error');
      } else {
        throw err;
      }
    }
  });

  /**
   * 6) Testing behavior when no request body is provided.
   */
  it('should return 400 or 422 if no request body is provided', async () => {
    const url = buildUrl(validProjectRef, validEnv, validName);

    try {
      await axios.put(url, {}, createAxiosConfig(authToken));
      throw new Error('Expected 400 or 422 error, but request succeeded.');
    } catch (err: unknown) {
      if (err instanceof AxiosError && err.response) {
        expect([400, 422]).toContain(err.response.status);
        expectJsonContentType(err.response);
        expect(err.response.data).toHaveProperty('error');
      } else {
        throw err;
      }
    }
  });

  /**
   * 7) (Optional) Testing rate limiting (429) if applicable.
   * This test is commented out by default because it might require multiple calls.
   * Uncomment if the API has rate-limit enforcement in place.
   */
  // it('should return 429 if the request is rate-limited', async () => {
  //   const url = buildUrl(validProjectRef, validEnv, validName);
  //   const requestData = { value: 'RateTest' };
  //
  //   for (let i = 0; i < 1000; i++) {
  //     try {
  //       await axios.put(url, requestData, createAxiosConfig(authToken));
  //     } catch (err: unknown) {
  //       if (err instanceof AxiosError && err.response) {
  //         if (err.response.status === 429) {
  //           expect(err.response.data).toHaveProperty('error');
  //           return;
  //         }
  //       }
  //     }
  //   }
  //   throw new Error('Expected 429 error, but request did not reach rate limit.');
  // });

  /**
   * 8) (Optional) Testing server errors (5xx). This is hard to force from the client side.
   * You could mock or intercept axios to simulate a 500 response.
   */
  // it('should handle 500 server error gracefully', async () => {
  //   // This test is typically done by mocking the API or error.
  // });
});

File: chapter_api_tests/2024-04/validation/test_put_api-v1-projects-{projectRef}-envvars-{env}-{name}.py
import axios, { AxiosResponse } from 'axios';
import { config as dotenvConfig } from 'dotenv';
import { describe, it, expect } from '@jest/globals';

dotenvConfig();

const API_BASE_URL = process.env.API_BASE_URL || 'http://localhost:3000';
const API_AUTH_TOKEN = process.env.API_AUTH_TOKEN || '';

/**
 * Jest test suite for PUT /api/v1/projects/{projectRef}/envvars/{env}/{name}
 *
 * Summary:
 * Update a specific environment variable for a specific project and environment.
 *
 * This suite tests:
 * 1. Input Validation (required params, body data types/edge cases).
 * 2. Response Validation (status codes, response body schema).
 * 3. Response Headers Validation (Content-Type and other headers).
 * 4. Edge & Limit cases (large payloads, empty values).
 * 5. Authorization & Authentication testing (valid, invalid, missing tokens).
 */

describe('PUT /api/v1/projects/{projectRef}/envvars/{env}/{name}', () => {
  const validProjectRef = 'my-project';
  const validEnv = 'staging';
  const validName = 'TEST_VARIABLE';

  // A guessed valid payload structure based on the assumption that the endpoint expects
  // a JSON body like: { value: string }
  const validPayload = {
    value: 'UPDATED_VALUE'
  };

  // Invalid payload - missing or incorrect fields
  const invalidPayload = {
    invalidKey: 123
  };

  // Axios default config with auth header
  const defaultAxiosConfig = {
    headers: {
      Authorization: `Bearer ${API_AUTH_TOKEN}`,
      'Content-Type': 'application/json'
    },
    validateStatus: () => true // Allow handling of non-2xx responses in tests
  };

  // Helper to build endpoint URL
  function buildUrl(projectRef: string, env: string, name: string): string {
    return `${API_BASE_URL}/api/v1/projects/${projectRef}/envvars/${env}/${name}`;
  }

  it('should update environment variable successfully with valid payload', async () => {
    const url = buildUrl(validProjectRef, validEnv, validName);

    const response: AxiosResponse = await axios.put(url, validPayload, defaultAxiosConfig);

    // Expect a 200 success response
    expect(response.status).toBe(200);

    // Validate response headers
    expect(response.headers['content-type']).toContain('application/json');

    // Validate response body (schema check can be more detailed if we have a schema)
    expect(response.data).toBeDefined();
    // Example: expect(response.data).toHaveProperty('message');
  });

  it('should return 400 or 422 for invalid body/payload', async () => {
    const url = buildUrl(validProjectRef, validEnv, validName);

    const response: AxiosResponse = await axios.put(url, invalidPayload, defaultAxiosConfig);

    // Expect a 400 or 422 for invalid request body
    expect([400, 422]).toContain(response.status);

    // Validate response headers
    expect(response.headers['content-type']).toContain('application/json');

    // Validate error response body
    expect(response.data).toBeDefined();
  });

  it('should return 400 if required path params are missing or invalid', async () => {
    // Attempt to call with invalid (empty) projectRef or env
    const url = buildUrl('', validEnv, validName);

    const response: AxiosResponse = await axios.put(url, validPayload, defaultAxiosConfig);

    // Expect a 400 or 404 here, depending on the API implementation
    // but typically a missing/invalid path param could lead to 400 or 404
    expect([400, 404]).toContain(response.status);
    expect(response.data).toBeDefined();
  });

  it('should return 401 or 403 for unauthorized/forbidden requests', async () => {
    // Remove the token from headers
    const noAuthConfig = {
      headers: {
        'Content-Type': 'application/json'
      },
      validateStatus: () => true
    };

    const url = buildUrl(validProjectRef, validEnv, validName);
    const response: AxiosResponse = await axios.put(url, validPayload, noAuthConfig);

    // Expect a 401 or 403 if no valid token is present
    expect([401, 403]).toContain(response.status);
    expect(response.data).toBeDefined();
    expect(response.headers['content-type']).toContain('application/json');
  });

  it('should return 404 if the resource is not found', async () => {
    // Use an obviously invalid environment variable name
    const url = buildUrl(validProjectRef, validEnv, 'NON_EXISTENT_VARIABLE');

    const response: AxiosResponse = await axios.put(url, validPayload, defaultAxiosConfig);

    // Expect a 404 if the environment variable does not exist
    expect(response.status).toBe(404);
    expect(response.headers['content-type']).toContain('application/json');
    expect(response.data).toBeDefined();
  });

  it('should handle large payload (edge case testing)', async () => {
    // Construct a large string
    const largeString = 'x'.repeat(5000);
    const largePayload = {
      value: largeString
    };

    const url = buildUrl(validProjectRef, validEnv, validName);
    const response: AxiosResponse = await axios.put(url, largePayload, defaultAxiosConfig);

    // Depending on implementation, can be success if large payload is allowed, or 400/413 if too large
    // Adjust expectations based on typical API behavior
    expect([200, 400, 413]).toContain(response.status);
    expect(response.headers['content-type']).toContain('application/json');
    expect(response.data).toBeDefined();
  });

  it('should return 400 or 422 when sending empty value in request body if not allowed', async () => {
    const emptyValuePayload = {
      value: ''
    };
    const url = buildUrl(validProjectRef, validEnv, validName);

    const response: AxiosResponse = await axios.put(url, emptyValuePayload, defaultAxiosConfig);

    // If the API does not allow empty strings for value, it might return 400 or 422
    // If the API does allow it, adjust the expectation accordingly
    expect([200, 400, 422]).toContain(response.status);
    expect(response.headers['content-type']).toContain('application/json');
    expect(response.data).toBeDefined();
  });
});
